backends/p4tools/common/compiler/compiler_target.h

```diff
@@ -25,28 +42,29 @@ class CompilerTarget : public Target {
     /// @returns any unprocessed arguments, or nullptr if there was an error.
     static std::vector<const char *> *initCompiler(int argc, char **argv);
 
-    /// Runs the P4 compiler to produce an IR.
+    /// Runs the P4 compiler to produce an IR and various other kinds of information on the input
+    /// program.
     ///
     /// @returns std::nullopt if an error occurs during compilation.
-    static std::optional<const IR::P4Program *> runCompiler();
+    static std::optional<const CompilerResult *> runCompiler();
```

Comment:
Is there any reason to return optional of a pointer? I don't see `std::nullopt` meaning something different than optional with `nullptr` (and if it did it would be quite confusing I think). I suggest using `std::optional<CompilerResult>` as the struct will likely mainly contain pointers/references and it will not be copied much around so this will matter even if it become larger.

Reply:
The pointer is necessary here because CompilerResult will be subclassed by the target back ends, e.g., BMv2CompilerResult etc. It could also be a reference, but that would require a reference_wrapper. 

I much prefer using `std::optional` over nullptr these days because it makes it very explicit that an error has happened. Going forward it will also be easier to replace this with std::expect or absl:StatusOr.

Reply:
hmm, I did not realize that `CompilerResult` will be subclassed, that does make a lot of sense. I don't like that `std::optional<pointer>` solution much because it has the ugly state where the optional contains `nullptr`. You are right that optional is more explicit about having to resolve the error though :-/ (and `std::optional<std::reference_wrapper<X>>` looks ugly and has the additional disadvantage that `auto x = opt.value()` behaves weirdly.

Reply:
I'm starting to think that `std::optional<std::reference_wrapper<const CompilerResult>>` is the right option here if you want the added safety of optional. This is a bit uglier to use (the type is long, and one needs to explicitly call `.get()` in many cases). But it is unlikely the `auto` problem I mentioned above will cause problems exactly because one has to call `get()` to get the actual instance. And having `std::optional<const CompilerResult *>(nullptr)` would be really confusing. Too bad std does not allow `std::optional<const CompilerResult &>` :-/.